### PR TITLE
YouTube updated to Version 3 of the API

### PR DESCRIPTION
**Pull request by Jarada**
This fixes Issue #174 by updating the YouTube module to the latest version 3 of Google's YouTube API.

Version 3 of the API requires the owner of the bot to provide a Google Developer Server API Key, and this is added as appropriate in the jenni core bot code for the default config file.

The functionality was kept as close as possible to the original module, while also including the .yt command for youtube_search (which is in active use in my community), and providing consistency with the rest of the modules by using jenni instead of bot.

### CODE REVIEW

**Comment by kaneda** [comment]
@Jarada can you add a piece in the readme about getting a Google dev API key to use this feature?

**Comment by Jarada** [comment]
Yup, will add that now :+1:

**Comment by Jarada** [comment]
I've added that above the Best Practices area. Let me know if that looks alright to you :)

File: modules/youtube.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
+#!/usr/bin/env python
+# vim: set fileencoding=UTF-8 :
+'''
+youtube.py - Youtube Module Improved v3
+
+Copyright 2015, Josh Begleiter (kanedasan@gmail.com)
```
Comment (kaneda) [inline modules/youtube.py]: You should probably give yourself some credit here :P

**Comment by kaneda** [comment]
@Jarada also feel free to add yourself here: https://github.com/myano/jenni/blob/master/CREDITS

**Comment by kaneda** [comment]
In general I think this is fine. @myano and I were trying to avoid requiring consumers of the bot to get an API key, but unfortunately, looking at the docs you provided in the issue, this may be unavoidable for YouTube. @myano any final thoughts here?

**Comment by Jarada** [comment]
Yeah, I don't like how Google have locked you in to using API keys to get this done now, the old way seemed to work well enough, but I suppose it helps with Google tracking who is doing what (and for managing quota's), so they get something out of it.

I'll add credit in, thanks! I do have some other pull requests for new features in the future, in specific, I have a variety of games similar to Uno to drop your way as well. That's for another day though.

**Comment by kaneda** [comment]
Sounds good. Something @myano and I had discussed was having a global entry point for tasks like this, but I worry about things like rate-limiting, DoS, etc.

**Comment by Jarada** [comment]
You mean a module that provides other modules access to the Google API to centralise that task?

**Comment by kaneda** [comment]
@Jarada i mean a proxy that we control that would connect to YouTube API through a single dev key

**Comment by Jarada** [comment]
@kaneda Ah, right. Yeah I would worry the same. It would be easy to abuse by anyone who got access to the central API Key, which would have to be public, not to mention that if all the users went over the quota, then none of the users have access. 

Providing individual API access offsets this risk to each consumer. It may be more painful but it ensures security of use and a large quota for each bot out there.

**Comment by kaneda** [comment]
@Jarada I agree. We could do it securely, but that isn't the issue so much as having to build in our own form of rate limiting, which could impact all consumers.

**Comment by kaneda** [comment]
@Jarada @myano is quite unhappy with the new YT API, I'm going to wait for his :+1: before revisiting this PR

**Comment by Jarada** [comment]
@kaneda I can understand that, given the idea of Jenni being as independent of external APIs as you want it to be. I'm in no rush to see this PR close out, just as and when you've figured out how you want to move Jenni forward with this. It's better not to rush changes that will change set dynamics without some thought; better to get it right then move quickly.

I do feel though that more services will be leaning the way of API keys in the future as they realise the power of using API keys to ensure compliance, tracking and analytics on the use of the services they provide. It is probably inevitable, so maybe the thing we should be looking into is not whether or not to include such services, but how to make it so that they are easier to manage for the end user. 

I'm not sure what the approach on that is; whether it's a guided setup provided by Jenni rather than entering it into the config file, or a webpage that explains how to go about setting these up.

Whatever you and @myano choose, make sure it feels right for you guys for the long term view of Jenni.

**Comment by kaneda** [comment]
Thanks for your patience @Jarada

**Comment by myano** [comment]
I love API's. In fact jenni uses and makes calls to countless APIs. What I do hate is when APIs require several keys to get minimum functionality for something basic. Twitter requires at least 2 API keys, and Google makes you create API keys for each specific service they offer rather than letting you one set of credentials.

I miss the days of simply using API keys in an old-fashioned GET request as part of the URL. Not this new fancy OAuth2 bullshit.

**Comment by kaneda** [comment]
@myano agreed, I think use auth where it makes sense, and fetching a YouTube title doesn't need auth, so why make us jump through hoops? I guess the question is, what is the future of this module and modules like it within Jenni. We can brainstorm offline in the room tonight if you have time.

**Comment by myano** [comment]
Maybe a YouTube module would be fine as a standalone feature that would be optional, instead of integration into the url.py where it would auto-YouTube URLs?

**Comment by kaneda** [comment]
@myano that sounds like a plan

**Comment by Jarada** [comment]
@kaneda @myano I have updated this PR to uncouple the YouTube module from the URL module. This makes the YouTube module a completely independent module, as suggested.

**Pull request merged**